### ISSUE TITLE
Standby stuck in startup when WAL not available

### ISSUE THREAD

**Issue by ants** [zalando/patroni#279]
When a server comes back after a crash and master has recycled the necessary WAL the standby can get stuck in startup. It's not possible to prevent this with replication slots because Patroni doesn't have a persistent concept of cluster members, but that's a separate issue. When this happens PostgreSQL will sit in a startup loop trying and failing to fetch WAL from the master, Patroni will fail to do anything because connection to PostgreSQL fails. Reinit and restart on the API just time out. Only way out is to stop Patroni and rm -rf data directory. It would be nice if it was possible to use Patroni commands to resolve this kind of situation, or even better, automatically reinit the the standby when this happens.

**Comment by ants** [comment]
I'll take a shot at a patch later tonight or tomorrow, but ideas on how to approach this would be welcome.

**Comment by CyberDem0n** [comment]
> Reinit and restart on the API just time out. 

It's kind of strange. 
Reinit is performed asynchronously and API should respond immediately that Reinitialize is either scheduled or started.

Restart in this case is useless but it should not time out either.

**Comment by ants** [comment]
The issue seems to stem from the fact that postgres process is running, but not accepting connections yet. I don't have the logs here on my laptop, but Patroni log was full of exceptions about PostgreSQL connections to load state information failing after timeout.

**Comment by bodymindarts** [comment]
We deal with this by running [this script](https://gist.github.com/bodymindarts/5dc15eaa409afd39e3621f887cc7343d) all the patroni containers.

**Comment by ants** [comment]
My current hypothesis is lack of fairness in threading.RLock. Ha.run_cycle() acquires async_executor for the whole duration of it's run and then blocks inside Postgresql.sync_replication_slots(). This eventually times out, but only after exceeding the cycle time, so Patroni.run() immediately calls run_cycle() again and it looks like some interaction with Python's GIL and threading module makes it so that other threads don't get a chance to acquire. [I wrote a short demo script that demonstrates the issue](https://gist.github.com/ants/c90ce6732c38fcdbf3c4a228180e5d39). This means that Ha.schedule() never gets a chance to run.

I'll now try to figure out the best way to fix this. Running sync_replication_slots asynchronously seems like an option.

**Comment by alexeyklyukin** [comment]
We tried to move the long-runnings ops (retrying on PostgreSQL request) out sync_replication_slots, please, here is the pull request: https://github.com/zalando/patroni/pull/282

**Comment by ants** [comment]
Tested #282, reinit now works correctly even when PG is stuck in startup.
